vignettes/creating-backend-classes.Rmd

```diff
@@ -0,0 +1,1413 @@
+#' List core chromatogram variables along with data types.
+coreChromVariables()
+```
+
+A typical `chromVariables()` method for a `ChromBackend` class will thus be
+implemented similarly to the one for our `ChromBackendTest` test backend: it
+will return the names for all available spectra variables within the backend
+object.
+
+```{r}
+#' Accessor for available chromatogram variables
+setMethod("chromVariables", "ChromBackendTest", function(object) {
+    names(object@chromData)
+})
+chromVariables(be)
+```
+
+
+### `chromData()`
+
+The `chromData` method should return the **full** chromatogram data within a
+backend as a `data.frame` object. A parameter `columns` should allow to define
+the names of the variables that should be returned. A parameter `drop` should
+also be implemented to allow for the calling of one column while still 
+controlling the return type.
+Each row in this data frame should represent one chromatogram, each column a 
+chromatogram variable. The `data.frame` **must** provide values (even if they
+are `NA`) for **all** requested chromatogram variables of the backend
+(**including** the core chromatogram variables). The `fillCoreChromVariables()` 
+function from the *Chromatograms* package allows to *complete* (fill) a
+provided `data.frame` with eventually missing core chromatogram variables:
+
+```{r}
+#' Get the data.frame with the available chrom variables
+be@chromData
+#' Complete this data.frame with missing core variables
+fillCoreChromVariables(be@chromData)
+```
+
+We can thus use this function to add eventually missing core chromatogram
+variables in the `chromData` implementation for our backend:
+
+```{r}
+#' function to extract the full chromData
+setMethod(
+    "chromData", "ChromBackendTest",
+    function(object, columns = union(names(coreChromVariables()),
+                                   chromVariables(object)),
+             drop = FALSE) {
+      if (!any(chromVariables(object) %in% columns) &&
+          !any(names(coreChromVariables()) %in% columns))
+        stop("Some of the requested Chromatogram variables are not ",
+                   "available")
+      if (all(columns %in% chromVariables(object)))
+        res <- object@chromData[, columns, drop = drop]
+      else {
+        res <- fillCoreChromVariables(object@chromData)
+        res <- res[, columns, drop = drop]
+        }
+      res
+      })
+```
+
+We can now use `chromData()` to either extract the full chromatogram data from
+the backend, or only the data for selected variables.
+
+```{r}
+#' Extract the full data
+chromData(be)
+
+#' Selected variables
+chromData(be, c("mz", "msLevel"))
+
+#' Only missing core spectra variables
+chromData(be, c("collisionEnergy", "mzMin"))
+```
+
+### `peaksVariables()`
+
+The `peaksVariables()` function is supposed to provide the names of the
+available *peaks variables*. Backends **must** provide retention time and
+intensity values, thus, the default implementation simply returns `c("rtime",
```

Comment:
mmm my thought process for now was to mainly develop the functions related to `chromData` and `peaksData` the same way. So if the default `chromVariables()` throws an error so does the default `peaksVariables()`. So that, as you say, the developers need to think about what to provide, and we offer flexibility. I will update the text here ! 

But do we want to ensure "rtime" and "intensity" in the test ? because then it clashes with the the idea above no... if we force it then might as well have a proper default implementation.. maybe we can discuss it more next time.

Reply:
yes - there is always a tradeoff between having defaults and forcing developers to do something. IMHO, `chromVariables()` should at least return (always) the `coreChromVariables()`, and `peaksVariables()` the `corePeaksVariables()`. we must ensure at some point that we get some data back from the backend - and for a chromatogram the minimal information we need is rtime and intensity. And for any of the access methods to work we need the codeChromVariables.

I think it would be good to test that these values are (minimally) returned by a backend in the unit test suite.

Reply:
@jorainer i was talking about this comment :) but you already answered aha

Reply:
Okay I understand it's fair. I kind of wanted to keep in flexible in case other type of chromatographic data to read using this package. but (at least for peaksData) Rtime and int will always be needed. I'll update the default function, UT and their respective documentation to make it clear

Reply:
Actually, for the chrom variables we would have the `fillChromData()` function that would take care of filling missing variables with NA. But a chromatogram without rtime and intensity does not make much sense, because then it's not a chromatogram. it can have additional peaks variable too, of course, but these two should IMHO be there.